Fix SIGSEGV in compress filter when appended before open

fdegros opened the pull request:
Bug: https://github.com/libarchive/libarchive/issues/2514
Test: libarchive_test test_read_append_compress_filter

mmatuska commented:
Please what is the purpose of this pull request? We already have a test_read_filter_compress.c and we currently don't test for individual archive_read_append_filter_*() functions.

If it would be desired, than all archive_read_filter_* tests should be extended with an additional test using archive_read_append_filter_*.

And to properly test the filter a compress compressed archive must be read.

In addition, you mix the contents of #2525 to this PR.

Tests fail.

kientzle commented:
All of the `test_read_append_*` tests in this file are pretty messed up, actually.  If someone is interested in improving these tests, I'd be happy to give some guidance.

kientzle commented:
Changed to "Draft" since this is still incomplete.

fdegros committed:
Fix SIGSEGV in compress filter when appended before open

Calling archive_read_append_filter(a, ARCHIVE_FILTER_COMPRESS) would
previously trigger a crash because compress_bidder_init() attempted to
read header bits from the upstream filter immediately. If the archive
was not yet opened (common when setting up filters), the upstream filter
state was not ready for reading.

This commit defers the header reading and decompressor initialization
until the first read operation (lazy initialization), consistent with
other filter implementations in libarchive.

Added test_read_append_compress_filter.

Bug: https://github.com/libarchive/libarchive/issues/2514
Test: ./libarchive_test test_read_append_compress_filter

fdegros committed:
Add tests for appending various filters before archive open

Extend test coverage to ensure all supported filters can be appended
to an archive reader before it is opened, matching the behavior
required to fix #2514.

Added tests for: BZIP2, LRZIP, LZ4, LZIP, LZMA, ZSTD, RPM, UU, and NONE.

kientzle reviewed (approved):
Nice!  Thanks for cleaning up these tests!
